### PR TITLE
(1) Add 3 tests for multi-datasets

### PR DESCRIPTION
**Pull request by vchoi-hdfgroup**
(2) Modify tests to make sure they cover type conversion for writing (3) Fail the checking of collective I/O mode

### CODE REVIEW

File: test/select_io_dset.c [hunk 1]
```diff
@@ -782,6 +820,665 @@ test_cmpd_with_bkg(hid_t fid, unsigned chunked)
...
+        }
+        else {
+            if (snprintf(dset_names[i], DSET_NAME_LEN,
+                         dtrans ? "multi_contig_trans_dset%u" : "multi_contig_ntrans_dset%u", i) < 0)
+                FAIL_STACK_ERROR;
+        }
+
+        /* dset0: H5T_NATIVE_INT, the other datasets: H5T_NATIVE_LLONG or H5T_NATIVE_SHORT */
+        if (i == 0) {
+            if ((dset_dids[i] = H5Dcreate2(fid, dset_names[i], H5T_NATIVE_INT, file_sids[i], H5P_DEFAULT,
```
Comment (fortnern) [inline test/select_io_dset.c]: I'd like to see both with and without type conversion for the first dataset. Also some of the other datasets should have no type conversion

File: test/select_io_dset.c [hunk 2]
```diff
@@ -782,6 +820,665 @@ test_cmpd_with_bkg(hid_t fid, unsigned chunked)
...
+                                           dcpl, H5P_DEFAULT)) < 0)
+                FAIL_STACK_ERROR;
+        }
+        else {
+            if ((dset_dids[i] =
+                     H5Dcreate2(fid, dset_names[i], ((HDrandom() % 2) ? H5T_NATIVE_LLONG : H5T_NATIVE_SHORT),
```
Comment (fortnern) [inline test/select_io_dset.c]: Maybe use int16_t, int32_t, and int64_t to ensure type conversions and size changes

File: test/select_io_dset.c (hunk 1 above)
Comment (vchoi-hdfgroup) [reply test/select_io_dset.c]: I think the first dataset (dset0) does have "with and without type conversions":

Before the 1st multi write/read, mem_tids[i] are set to native int:

for (i = 0; i < (int)ndsets; i++)
        mem_tids[i] = H5T_NATIVE_INT;

And before the 2nd multi-read, mem_tids[i] are set to native long:

for (i = 0; i < (int)ndsets; i++)
        mem_tids[i] = H5T_NATIVE_LONG;

File: test/select_io_dset.c [hunk 3]
```diff
@@ -782,6 +820,665 @@ test_cmpd_with_bkg(hid_t fid, unsigned chunked)
...
+        if (i == 0 || i == mm)
+            continue;
+        if (H5Sselect_none(mem_sids[i]) < 0)
+            TEST_ERROR;
+        if (H5Sselect_none(file_sids[i]) < 0)
+            TEST_ERROR;
+    }
+
+    /* Datatype setting for read from <mm> dataset */
+    mem_tids[mm] = ss_bc_tid;
```
Comment (fortnern) [inline test/select_io_dset.c:1369]: This line seems to be repeated unnecessarily

File: test/select_io_dset.c (hunk 2 above)
Comment (vchoi-hdfgroup) [reply test/select_io_dset.c]: I will add another test for size changes with H5T_STD_I16BE/H5T_STD_I32BE/H5T_STD_I64BE etc.

File: test/select_io_dset.c (hunk 1 above)
Comment (vchoi-hdfgroup) [reply test/select_io_dset.c]: I will modify this test to be:
--Create datasets with randomized H5T_NATIVE_INT or H5T_NATIVE_LONG
--Multi write/read with H5T_NATIVE_INT
--Multi write/read with H5T_NATIVE_LONG

File: test/select_io_dset.c (hunk 3 above)
Comment (vchoi-hdfgroup) [reply test/select_io_dset.c:1369]: Removed.

**Review by fortnern — APPROVED** [review]

**Pull request merged**